mrmustard/utils/serialize.py

```diff
@@ -0,0 +1,121 @@
+
+    This function can serialize any object that implements the MrMustard
+    serialization interface, which is the following:
+
+    .. code-block:: python
+
+        @classmethod
+        def deserialize(cls, data: Dict) -> cls
+
+    Though not required to satisfy the interface, it is conventional for a
+    class to implement an instance method called ``serialize`` that takes no
+    arguments, calls this function with its data, and returns the Path result.
+
+    Args:
+        cls (type): The object type being serialized
+
+    Kwargs:
+        arrays (Dict[str, NDArray]): A dict of array-like objects to be serialized
+        data (Dict[Any, Any]): A JSON-serializable dict of arbitrary data to save
+
+    Returns:
+        Path: the path to the saved object, to be retrieved later with ``load``
+    """
+    file = CACHE / f"{cls.__qualname__}_{uuid4().hex}.json"  # random filename
+    data["class"] = f"{cls.__module__}.{cls.__qualname__}"
+
+    if arrays:
+        if overlap := set(arrays).intersection(set(data)):
+            raise ValueError(f"Arrays cannot have the same name as generic data: {overlap}")
+        backend = math.backend_name
+        npz = file.with_suffix(".npz")
+        np.savez(npz, **{k: math.asnumpy(v) for k, v in arrays.items()})
+        data["arrays"] = str(npz)
+        data["backend"] = backend
+
+    with file.open("w", encoding="utf-8") as f:
+        json.dump(data, f)
+
+    return file
+
+
+def load(file: Path, remove_after=True):
+    r"""
+    The deserializer entrypoint for objects saved with the ``save`` method.
+
+    Args:
+        file (Path): The json file to load from
+        remove_after (Optional[bool]): Once load is complete, delete the saved file
+    """
+    file = Path(file)
+    with file.open("r", encoding="utf-8") as f:
+        data = json.load(f)
+
+    if "arrays" in data:
+        npz_file = data.pop("arrays")
+        if (backend := data.pop("backend")) != math.backend_name:
+            raise TypeError(
+                f"Data serialized with {backend} backend, cannot deserialize to the currently active {math.backend_name} backend"
+            )
+        data.update(**{k: math.astensor(v) for k, v in np.load(npz_file).items()})
+        if remove_after:
+            Path(npz_file).unlink()
+
+    if remove_after:
+        file.unlink()
+    return locate(data.pop("class")).deserialize(data)
+
+
+def get_zipfile(name: str = None) -> Path:
+    r"""
+    Returns a randomly named zipfile in the cache folder. If a name is
+    provided, returns a Path for that base name in the cache folder.
+    """
+    return CACHE / (name or f"collection_{uuid4().hex}.zip")
+
+
+@contextmanager
+def cache_subdir(name=None):
+    r"""Context manager to have calls to ``save`` write to a cache subdirectory."""
+    global CACHE  # pylint:disable=global-statement
+    if not name:
+        name = f"subcache_{uuid4().hex}"
+
+    old_cache = CACHE
+    CACHE /= name
+    CACHE.mkdir()
```

Comment:
```suggestion
    CACHE.mkdir(parents=True)
```
seems the `".serialize_cache"` won't be there if subdir is used the first time serialization is used.

Reply:
it will be there because I'm adding `/.serialize_cache/.gitkeep` in this PR 😄